synapse/util/task_scheduler.py

```diff
@@ -0,0 +1,244 @@
+
+    async def get_tasks(
+        self, action: str, resource_id: Optional[str]
+    ) -> List[ScheduledTask]:
+        """Get a list of tasks associated with an action name, and
+        optionally with a resource id.
+
+        Args:
+            action: the action name of the tasks to retrieve
+            resource_id: if `None`, returns all associated tasks for
+                the specified action name, regardless of the resource id
+
+        Returns: a list of `ScheduledTask`
+        """
+        return await self.store.get_scheduled_tasks(action, resource_id)
+
+    async def _scheduled_tasks_loop(self) -> None:
+        """Main loop taking care of launching the scheduled tasks when needed."""
+        for task in await self.store.get_scheduled_tasks():
+            if task.id not in self.running_tasks:
+                if (
+                    task.status == TaskStatus.SCHEDULED
+                    and task.timestamp < self.clock.time_msec()
+                ):
+                    await self._launch_task(task, True)
+                elif task.status == TaskStatus.ACTIVE:
+                    await self._launch_task(task, False)
+                elif (
+                    task.status == TaskStatus.COMPLETE
+                    or task.status == TaskStatus.FAILED
+                ) and self.clock.time_msec() > task.timestamp + TaskScheduler.KEEP_TASKS_FOR_MS:
+                    await self.store.delete_scheduled_task(task.id)
+
+    async def _launch_task(self, task: ScheduledTask, first_launch: bool) -> None:
+        """Launch a scheduled task now.
+
+        Args:
+            task: the task to launch
+            first_launch: `True` if it's the first time is launched, `False` otherwise
+        """
+        if task.action not in self.actions:
+            raise Exception(
+                f"No function associated with the action {task.action} of the scheduled task"
+            )
+
+        function = self.actions[task.action]
+
+        async def wrapper() -> None:
+            try:
+                (status, result, error) = await function(task, first_launch)
+            except Exception:
+                f = Failure()
+                logger.error(
+                    f"scheduled task {task.id} failed",
+                    exc_info=(f.type, f.value, f.getTracebackObject()),
+                )
+                status = TaskStatus.FAILED
+                result = None
+                error = f.getErrorMessage()
+
+            await self.update_task(
+                task.id,
+                status=status,
+                result=result,
+                error=error,
+            )
+            self.running_tasks.remove(task.id)
+
+        await self.update_task(task.id, status=TaskStatus.ACTIVE)
+        self.running_tasks.add(task.id)
+        description = task.action
+        if task.resource_id:
+            description += f"-{task.resource_id}"
+        run_as_background_process(description, wrapper)
```

Comment:
Does it matter that this might spin up tons and tons of tasks? I kind of assumed this was only going to allow a certain number of active tasks at once (kind of like what we do with background updates).

Can a task purposefully stop but get re-run in the future? (E.g. if I'm a really big task but I want to only do some of the work now, then do more later. Kind of like we do with the background updates?)

Reply:
> Does it matter that this might spin up tons and tons of tasks? I kind of assumed this was only going to allow a certain number of active tasks at once (kind of like what we do with background updates).

Yeah I should probably add the 2 needed lines to limit the number of in flight tasks. Do you have an opinion on what would be the limit ? 5, 10, 50 ? I would say 10 is probably fine ?

> 
> Can a task purposefully stop but get re-run in the future? (E.g. if I'm a really big task but I want to only do some of the work now, then do more later. Kind of like we do with the background updates?)

There is one parameter missing in `update_task` to support this use case which I m going to add, currently `timestamp` is updated to now when running `update_task`, but if we take an optional parameter instead we can just put back the task in `SCHEDULED` status with a new timestamp in the future.
You can also just keep the task `ACTIVE`, in this case it will be run again on the next scheduler loop (if the number of active task is not too big already, considering we want to limit those).

Reply:
I think an arbitrary number (10?) would be fine.

Reply:
I put 20.